WcaOnRails/db/migrate/20180201005000_add_round_results_to_rounds.rb

```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class AddRoundResultsToRounds < ActiveRecord::Migration[5.1]
+  def change
+    add_column :rounds, :round_results, :text
```

Comment:
I like how simple this is. As I understand it, there are two issues to doing things this way:

1. Now every time we load a round via ActiveRecord, we also load it's `round_results` column. If these columns get large, and we do this a lot, that could have a performance cost. One way around that would be to put the round results in their own table, so we only load them when we need them. For the record, I don't think we actually need to do that (feels like premature optimization to me), but it's just something to keep in mind.
2. Have you tested how large this can get? Even if MySQL has a very large limit for `text` type, there's also the question of if we hit any limitations in ActiveRecord (something I believe @SqAtx ran into while working on the results submission form). I think it would be good to manually test out your API with a very large round just to confirm that everything works.

Reply:
1. That's a valid point. On the other hand, do we use rounds for something else than dealing with WCIF?
2. Thanks for pointing this out! You are surely right, `text` is not the appropriate type here. Just have a look at this super simple approximation:
```ruby
thousand_results = [{
  "personId" => 100,
  "ranking" => 100,
  "attempts" => [{ "result" => 4560, "reconstruction" => nil }] * 5,
}] * 1000

JSON.dump(thousand_results).length
```
```ruby
=> 234001
```

`TEXT` can store up to `65,535` characters, so in this case we would exceed the limit ~3.6 times.

A reasonable solution would be to use `MEDIUMTEXT` instead. There's even `LONGTEXT`, but I think it would be an overkill. `MEDIUMTEXT` stores up to `16,777,215` characters, so it could store around *6000* such results, given each attempt would have a reconstruction consisting of *500* characters (not sure which one is more hilarious =)).

So, unless you see anything wrong with the above reasoning, I'd just go for `MEDIUMTEXT`.

Reply:
1. Not that I know of, so I agree, we're fine here for now =)
2. MEDIUMTEXT sounds good to me too. Can you confirm that AR throws an exception if we try to put too long of a string into a column? I don't want us to be setting ourselves up for any issues like https://github.com/thewca/worldcubeassociation.org/pull/2161.

Reply:
That was not so easy to reproduce. I tried two things:

A. Rails console
```bash
bin/rails c --sandbox
```
```ruby
ten_thousand_results = [{
  "personId" => 100,
  "ranking" => 100,
  "attempts" => [{ "result" => 4560, "reconstruction" => "R'" * 150 }] * 5,
}] * 10_000; nil

JSON.dump(ten_thousand_results).length
#=> 17240001
# MEDIUMTEXT stores up to 16,777,215 (< 17,240,001)

round = Round.first
(round.round_results = RoundResults.load(ten_thousand_results)); nil
#=> (0.4ms)  ROLLBACK TO SAVEPOINT active_record_1
#=> ActiveRecord::StatementInvalid: Mysql2::Error: MySQL client is not connected: ROLLBACK TO SAVEPOINT active_record_1
```

It looks like database connection died because of this.

B. MySQL directly
```bash

printf "UPDATE rounds SET round_results = '$(printf "C%.0s" {1..17000000})' LIMIT 1;" > tmp/mysql_query.sql
mysql -u root -p wca_development
# MySQL cli
mysql> SET GLOBAL max_allowed_packet = 20000000; -- A small value would prevent form the update anyway.
mysql> source tmp/mysql_query.sql
Query OK, 0 rows affected, 1 warning (0.50 sec)
Rows matched: 1  Changed: 0  Warnings: 1
mysql> SHOW WARNINGS;
+---------+------+----------------------------------------------------+
| Level   | Code | Message                                            |
+---------+------+----------------------------------------------------+
| Warning | 1265 | Data truncated for column 'round_results' at row 1 |
+---------+------+----------------------------------------------------+
1 row in set (0.00 sec)
```

Looks fine I'd say. And I really see no way for running out of space using `MEDIUMTEXT`.

Reply:
Nice! And you've confirmed that the database connection does not die when you have, say 5 thousand results? As I recall, @SqAtx was trying to insert an amount of text that the MySQL column could fit, but AR was dying for some reason.

Reply:
:shipit: 

```bash
bin/rails c --sandbox
```

```ruby
five_thousand_results = [{
  "personId" => 100,
  "ranking" => 100,
  "attempts" => [{ "result" => 4560, "reconstruction" => "R'" * 150 }] * 5,
}] * 5_000; nil

JSON.dump(five_thousand_results).length
#=> 8620001
# MEDIUMTEXT stores up to 16,777,215 (> 8,620,001)

round = Round.first
(round.round_results = RoundResults.load(five_thousand_results)); nil
# (0.2ms)  RELEASE SAVEPOINT active_record_1
#=> true
```

Reply:
Woah, cool! I did not know about `--sandbox`, that's terribly useful =)

:shipit:!